siobrultech_protocols/gem/protocol.py

```diff
@@ -120,5 +133,73 @@ def skip_malformed_packet(msg, *args, **kwargs):
             except MalformedPacketException as e:
                 skip_malformed_packet(e.args[0])
 
+        self._ensure_transport()
+
+    def _ensure_transport(self) -> asyncio.BaseTransport:
         if not self._transport:
             raise EOFError
+
+        return self._transport
+
+
+class BidirectionalProtocol(PacketProtocol):
```

Comment:
In hindsight, I realize that this doesn't meat the [sans-I/O philosophy](https://sans-io.readthedocs.io/) this repo is supposedly supposed to be about.  Do we:
a) not care
b) spin this protocol out into a new library (`brultech_protocols`?)
c) rename this, and deprecate the existing `siobrultech_protocols`

Reply:
Another one that is less painful
(d) rename this repository to `brultech_protocols`, continue to have `siobrultech_protocols` that is the Sans-I/O library, and the new library that we'd have to come up with a name for.

(a) and (b) are the least amount of work

Reply:
Went and read that and some other things about it. Sounds like there's a bit of a learning curve and a lot of subtleties to that philosophy that I'm just beginning to learn, but here are some thoughts:

As I understand it, there are two things that are problematic about this from a sans-I/O standpoint:
1. It implements `asyncio.Protocol` and uses `asyncio.Queue` and `asyncio.Transport`, and thus has a dependency on a specific I/O library. Personally I think we can let this one slide with the notion that what we have is easy to integrate with any I/O library (which is the point) and it's *really* easy to integrate with `asyncio`, but I see reasonable counter-arguments to that. Reasons why it's easy to integrate with any I/O library:
    * `Protocol` is an interface of more or less the right shape for sans-I/O
    * `Transport` is effectively using a callback to return bytes, which is not a return value or a queryable internal buffer as suggested by sans-I/O, but can trivially be transformed into the latter
    * `Queue` can be accessed in a non-`async` way and thus is more-or-less equivalent to using a `list` or `deque`

2. It's using coroutines and locking. I can rework this in a way that doesn't do that. Something like:
    * provide a synchronous `begin_api_request` method that passes the PDL command bytes to the `Transport` and returns something that indicates how long to wait before sending the API request
    * provide synchronous `send_X_request` methods that switch the protocol into api mode, pass the appropriate request bytes to the `Transport`, and return something that indicates how long to wait before requesting the API response
    * provide synchronous `get_X_response` methods that parse the appropriate API response out of the internal buffer
    * provide a synchronous `end_api_request` method that clears the internal buffer and switches back to packet mode
    * or maybe the above two are actually a collection of classes, one per API call, with `send_request` and `get_response` methods on them
    * we'd keep track of state and the methods above would get mad if they're called in the wrong order

In `greeneye-monitor` I'd wrap all those synchronous method calls in `async` ones that take a lock and invoke all that stuff with appropriate waiting and a context manager for ending API mode if anything blows up in the middle.

Reply:
I agree on (1).

> or maybe the above two are actually a collection of classes, one per API call, with send_request and get_response methods on them

We might be setting ourselves up to write a lot of code here.  I'm not opposed, but pondering if it's worth it. These projects are both for learning and accomplishing energy monitoring for me, so I'm okay with it, but want to make sure I'm not making this harder than you want it to be.

> In greeneye-monitor I'd wrap all those synchronous method calls in async ones that take a lock and invoke all that stuff with appropriate waiting and a context manager for ending API mode if anything blows up in the middle.

Myself and others would probably find that useful, so if that's the path you want to go down, I'd love to see it in an intermediate library.

Reply:
Re difficulty, I think it would likely end up looking like the use of `PacketFormat` today: one class, constructed in a variety of ways.

If you think there's room for another library between this one and the higher-level API provided by `greeneye-monitor`, I'm not opposed. Maybe `aiobrultech-protocols`? I don't think there's gonna be enough code in it to be worth a whole new GitHub project and the attendant overhead (vs. publishing two libraries out of this one), but up to you.

So for this PR I'll rework this to be more sans-I/O-like. I'll do the `asyncio` piece in `greeneye-monitor` for now since I'd like to get back to what I'm working on there, and whenever you get to a point in your project where you want it, it'll be easy to extract to another library (or use directly while ignoring the higher level stuff from that library).

Reply:
I'm going on a 30-day sabbatical from work starting on Saturday, and one of the things I plan to work on during that time is getting energy monitoring working in Home Assistant, so I'll be dedicating a bunch of time to close this out soon :)

Reply:
I think we can resolve this thread now, yeah?